util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,266 @@
+		return
+	}
+	klog.V(3).Infof("Marshalled Response: %v", response)
+	rw.WriteHeader(statusCode)
+	rw.Write(marshalledResponse)
+}
+
+// StartHTTPServer starts an http server for http measurements
+func (w *Worker) StartHTTPServer(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting HTTP Server")
+	go w.startListening(httpPort, handlersMap{"/test": w.Handler})
+	w.sendResponse(rw, http.StatusOK, nil)
+}
+
+// StartTCPServer starts iperf server for tcp measurements
+func (w *Worker) StartTCPServer(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting TCP Server")
+	w.startWork(rw, request, ProtocolTCP, "iperf", tcpServerArguments)
+}
+
+// StartUDPServer starts iperf server for udp measurements
+func (w *Worker) StartUDPServer(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting UDP Server")
+	w.startWork(rw, request, ProtocolUDP, "iperf", udpServerArguments)
+}
+
+// StartHTTPClient starts an http client for http measurements
+func (w *Worker) StartHTTPClient(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting HTTP Client")
+	w.startWork(rw, request, ProtocolHTTP, "siege", httpClientArguments)
+}
+
+// StartTCPClient starts iperf client for tcp measurements
+func (w *Worker) StartTCPClient(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting TCP Client")
+	w.startWork(rw, request, ProtocolTCP, "iperf", tcpClientArguments)
+}
+
+// StartUDPClient starts iperf client for udp measurements
+func (w *Worker) StartUDPClient(rw http.ResponseWriter, request *http.Request) {
+	klog.Info("Starting UDP Client")
+	w.startWork(rw, request, ProtocolUDP, "iperf", udpClientArguments)
+}
+
+// Handler handles http requests for http measurements
+func (w *Worker) Handler(rw http.ResponseWriter, request *http.Request) {
+	w.sendResponse(rw, http.StatusOK, "hi")
+}
+
+func (w *Worker) startWork(rw http.ResponseWriter, request *http.Request, protocol string, command string, arguments []string) {
+	arguments, err := populateTemplates(arguments, request)
+	if err != nil {
+		w.sendResponse(rw, http.StatusBadRequest, status{Message: err.Error()})
+		return
+	}
+	valMap, err := getValuesFromURL(request, []string{"timestamp", "duration"})
+	if err != nil {
+		w.sendResponse(rw, http.StatusBadRequest, status{Message: err.Error()})
+		return
+	}
+	timestamp, err := strconv.ParseInt(valMap["timestamp"], 10, 64)
+	if err != nil {
+		klog.Infof("Invalid timestamp: %v %v", valMap["timestamp"], err)
+	}
+	go w.schedule(timestamp, valMap["duration"], protocol, command, arguments)
+	w.sendResponse(rw, http.StatusOK, nil)
+}
+
+func (w *Worker) schedule(startTimestamp int64, duration string, protocol string, command string, arguments []string) {
+	defer close(w.stopCh)
+	w.startTime = time.Unix(startTimestamp, 0)
+	klog.Infof("About to wait until %v, current time: %v", w.startTime, time.Now())
+	time.Sleep(w.startTime.Sub(time.Now()))
+	w.workerDelay = time.Now().Sub(w.startTime)
```

Comment:
merge into single log

w.startTime := time.Unix(startTime, 0)
now := time.Now()
w.workerDelay = now.Sub(w.startTime)
klog.Infof("About to wait until %v, current time: %v", w.startTime, now)
time.Sleep(w.workerDelay)
 
@VivekThrivikraman-est VivekThrivikraman-est 2 hours ago 
Have merged into a single log, but there are a couple of issues with the pseudocode.
1)If we calculate delay before sleep,the workerDelay can have negative values when startTime > now, if it was after sleep delay would have be almost 0(which I think would be more appropriate).
2)time.Sleep(w.workerDelay) will incorrectly sleep as sleep should be for (startTime-now) whereas workerDelay is (now-startTime) , or we would have to do time.Sleep(w.workerDelay * -1)

Please let me know your opinion.